### ISSUE TITLE
OutOfMemory Error when creating large worlds with IncDegreeStrategy

### ISSUE THREAD

**Issue by scutze** [tristndev/ComPI#2]
See title. Larger worlds also seem to take exponentially longer to create. Stack Trace:

![image](https://user-images.githubusercontent.com/6710211/107051215-79d62a00-67cc-11eb-9e40-661b49f91a03.png)

If [Line 75](https://github.com/tristndev/ComPI/blob/b75acf8e3da364cc2ebb05c8dfa2ea4eea5a2d5c/1_BLOGBuilder/src/factories/augmentation/IncDegreeFactory.java#L75) in the factory is changed to [Line 69](https://github.com/tristndev/ComPI/blob/b75acf8e3da364cc2ebb05c8dfa2ea4eea5a2d5c/1_BLOGBuilder/src/factories/augmentation/IncByWorldFactory.java#L69) of the IncWorld Factory, where the merge point is selected randomly, the symptoms disappear. 

##### Steps to reproduce:
 * choose IncDegreeStrategy
 * set incrementCount to e.g. 100
 * Run.

**Comment by tristndev** [comment]
When I ran your strategy I also already noticed that it took longer and longer for the increased augmentation steps but didn't put more thought into it. 

A possible idea what could be the issue (even though I am not sure yet how that could explain the exponential time increases):
You set the `mergePoint` reference with your initial model. Then the model is cloned and cloned and cloned with each augmentation step. Maybe the `mergePoint` reference is then outdated after the first clone because it points towards the initial model?

Just a first idea. But I will try to take a look at it later. 👍🏼 

--- 

*Edit:* 

I think this really is the problem. Maybe we can fix it by writing a `getRandVarByIndex(index)` method in the [`World.java`](https://github.com/tristndev/ComPI/blob/b75acf8e3da364cc2ebb05c8dfa2ea4eea5a2d5c/1_BLOGBuilder/src/blogbuilder/World.java#L610)? Then you would just save the randvar index of the initial merge point and could get the reference to the corresponding randvar in the current clone of the world.

Alternatively, try out as your [Line 75](https://github.com/tristndev/ComPI/blob/b75acf8e3da364cc2ebb05c8dfa2ea4eea5a2d5c/1_BLOGBuilder/src/factories/augmentation/IncDegreeFactory.java#L75) with the index to the merge-randvar as `mergePointIndex`:
```java
RandVar rndRandVarW1 = RandVar rndRandVarW1 = w1.getRandVars().get(mergePointIndex);
```
This might also work since the order of the randvars is never changed, I think. ✊🏼
